consensus/src/epoch_manager.rs

```diff
@@ -304,9 +304,25 @@ impl EpochManager {
 
         let mut safety_rules =
             MetricsSafetyRules::new(self.safety_rules_manager.client(), self.storage.clone());
-        safety_rules
-            .perform_initialize()
-            .expect("Unable to initialize SafetyRules");
+        match safety_rules.perform_initialize() {
+            Ok(()) => { /* Expected */ }
+            Err(Error::ValidatorKeyNotFound(error)) => {
+                error!(
+                    error = error,
+                    "Unable to find the validator consensus key during initialize",
```

Comment:
It's unclear to me what happens in this case.  Should the operator just let the validator keep running and then it might catch up to the point where the history matches the keys in vault and then restart the validator?

Reply:
It's basically the same as sync_only mode. It doesn't require restart since it tries to initialize every epoch.

Reply:
Yes, so the current plan is to add an alert for this error, so that when it fires, operators will be aware and manually inspect the error. Upon manual inspection, there's two cases:
1. Assuming the validator is simply behind the blockchain head, the operator will let the validator keep running until it catches up with the current head of the blockchain, at which point LSR will be able to get the keys from storage and things will continue as normal (no restart required).
2. If something is very broken (e.g., secure storage is dead), the operator will need to escalate and try to figure out what's gone wrong.

From previous experience, most operators who run into 1 are already experiencing problems (e.g., they've tried to migrate a node across operators, failed in doing an upgrade, or seen a previous failure which meant their node is significantly behind).

'cc @davidiw who is going to be adding the alerts for this change.

Reply:
Thanks for the explanation.

Regarding 1,  we could use the current time compared to the blockchain head time to give a hint at least of how far behind it is - e.g. it's more of an info if it's just really out-of-date compared to being an error.  Also, it might be nice to add a comment that this is a potentially transient error....now I'm starting to wonder if this should be warn rather than error since it should self-heal.

Regarding 2, if secure storage is dead, shouldn't we just propagate that error rather than ValidatorKeyNotFound?

Reply:
This should simply be an Error that SafetyRules failed to be initialized as there can be lots of reasons why it didn't.

Reply:
My concern is that an operator that sees this error won't understand that on a subsequent epoch that initialize might succeed from this error message.  Error generally means something bad happened and the operator might need to take action, but I don't see how this message would communicate what needs to be done or not done.  It would probably just be escalated to a SME in this area when it doesn't need to be with more clear documentation.  I don't mind more verbose error messages as they are supposed to be quite uncommon if we can help operators understand what is going on and what actions to take (even if it's just pointing at other documentation).

Reply:
> My concern is that an operator that sees this error won't understand that on a subsequent epoch that initialize might succeed from this error message... 

My general opinion here is that the alert triggered for this error should contain all the information for the operator to react to it, i.e., they won't really know if an error is logged in LSR, unless an alert fires, right? (Operators won't be tailing the logs in practice 😄) In which case I was envisioning us having the alert that triggers here point directly to the runbook, which should entail exactly how to react to the alert. Note: I don't mind having verbose error logs, but I was always under the impression that our runbooks were the place for this information to be (hence why we link to runbooks in alerts?)

> Regarding 1, we could use the current time compared to the blockchain head time to give a hint at least of how far behind it is..

This would be great, but it's difficult to identify how far we are behind the head (without having another node to query).. Hence why, it requires some sort of human reasoning. Unless we think it's practical to be pinging other validators for this information? (I'm personally not in favor of this)

> now I'm starting to wonder if this should be warn rather than error since it should self-heal.

I think this can correctly be dealt with at the alert level, e.g., if the alert fires for a few minutes, issue a warning (it could still be catching up, and the runbook should tell the operator that it might fix itself if the operator believes the node to be behind). If the alert continues to fire for 30 minutes+, then increase the alert warning and say that they're unlikely to catch up and it requires intervention?

Reply:
Note: once we've settled on the right approach that makes everyone happy, I'll definitely follow up with the improvements/fixes we want for these changes. Next time I also won't be so fast as to land soon after an approval 😄

Reply:
I agree on the point about notification via an alarm about an issue and a runbook is a more detailed place to describe the process of resolving it.  That being said, it would be helpful for an error to convey the seriousness of the issue and implications of it from a standalone point of view of someone who views it.  

By the way, this PR definitely is an improvement, I'm just thinking what else we could do to make it clearer to someone who sees this message.  If I saw it, the first thing I would do is start reading the code trying to extract what to do next.  Personally, I'd prefer to see something more like "Unable to find the validator consensus key during initialize() for epoch XXX and will not participate in consensus.  Currently, version=YYY, blockchain time=ZZZ, epoch=WWW, and will continue syncing until the validator consensus key matches for a future epoch."

Reply:
Okay, thanks @aching! I can see the benefit in conveying the urgency/context a little more clearly, like you say. Although, I do have some concerns returning that exact error message, but let's debate it in a follow up PR that cleans things up here 😄